source/grid/tria_accessor.cc

```diff
@@ -1329,139 +1329,53 @@ namespace
-                               0.86595709258347853 * c +
-                               1.3399810435848563 * d + 1.8611363115940525 * e +
-                               1.2469436885317342 * f) +
-                   0.03025074832140047 * std::sqrt(a + 0.86595709258347853 * b +
-                                                   0.0048207809894260144 * c +
-                                                   1.8611363115940525 * d +
-                                                   0.13886368840594743 * e +
-                                                   0.12922212642709538 * f) +
-                   0.056712962962962937 *
-                     std::sqrt(
-                       a + 0.86595709258347853 * b + 0.10890625570683385 * c +
-                       1.8611363115940525 * d + 0.66001895641514374 * e +
-                       0.61419262306231814 * f) +
-                   0.056712962962962937 *
-                     std::sqrt(a + 0.86595709258347853 * b +
-                               0.44888729929169013 * c +
-                               1.8611363115940525 * d + 1.3399810435848563 * e +
-                               1.2469436885317342 * f) +
-                   0.03025074832140047 *
-                     std::sqrt(a + 0.86595709258347853 * b +
-                               0.86595709258347853 * c +
-                               1.8611363115940525 * d + 1.8611363115940525 * e +
-                               1.7319141851669571 * f);
-          }
-
-        // the face is planar. then its area is 1/2 of the norm of the
-        // cross product of the two diagonals
-        const Tensor<1, 3> v12        = accessor.vertex(2) - accessor.vertex(1);
-        const Tensor<1, 3> twice_area = cross_product_3d(v03, v12);
-        return 0.5 * twice_area.norm();
+        const Point<3> x0 = accessor.vertex(0), x1 = accessor.vertex(1),
+                       x2 = accessor.vertex(2), x3 = accessor.vertex(3);
+
+        // This is based on the approach used in libMesh (see face_quad4.C): the
+        // primary differences are the vertex numbering and quadrature order.
+        //
+        // The area of a surface is the integral of the magnitude of its normal
+        // vector, which may be computed via the cross product of two tangent
+        // vectors. We can easily get tangent vectors from the surface
+        // parameterization. Hence, given a bilinear surface
+        //
+        //     P(chi, eta) = x0 + (x1 - x0) chi + (x2 - x0) eta
+        //                 + (x3 + x0 - x1 - x2) chi eta
+        //
+        // the tangent vectors are
+        //
+        //     t1 = (x1 - x0) + (x3 + x0 - x1 - x2) eta
+        //     t2 = (x2 - x0) + (x3 + x0 - x1 - x2) xi
+        const Tensor<1, 3> a = (x3 + x0 - x1 - x2), b0 = x1 - x0, b1 = x2 - x0;
+
+        // If we have a parallelogram then return quickly.
+        if (a.norm_square() < 1e-14 * accessor.diameter())
+          return cross_product_3d(b0, b1).norm();
```

Comment:
Since you use `norm_square`, with which I agree, shouldn't you also use the square of the diameter or some other measure like `b0 * b0` to have the same units?

Reply:
Probably... I think the answer is to square the diameter but I'll double-check.

Reply:
There's probably a fancier way to do this, but since `a` is the distance of the last vertex from the parallelogram defined by the first three vertices, I think its reasonable to just check that its components are near zero. Hence I switched this to use an infinity norm for `a`.